Add dismiss handle to Information Sheet

charliescheer opened the pull request:
### Fix
This PR continues with the design changes in issue #1061. Specifically this PR adds in a visual element to alert the user that they can dismiss the information sheet by dragging down on the top of the information sheet element.

The drag bar element will not appear on iPad and should look a bit like this:

<img src="https://cdn-std.droplr.net/files/acc_593859/WpyBbZ" />

### Test

1. Open up Simplenote on an iPhone and go to a Note > Information sheet, confirm that the drag bar appears
2. Open up Simplenote again on an iPad and go to a note > Information sheet, confirm there is no drag bar

### Review
One developer and one designer are required to review these changes, but anyone can perform the review.

### Release
> These changes do not require release notes.

charliescheer committed:
Added drag bar to Information sheet

charliescheer committed:
Changed dragBar to not appear if device is iPad

charliescheer commented:
I pulled this PR out of the other one I made for the information sheet style updates because I wanted to ask if this is something that we want to extract into it's own class.  It is rather simple, just a small UIView with a background color, but if it turned out to be something that we wanted to insert onto other screens, I could see how it could be setup to simply be added to any modal and save on repeat code later. 

Thoughts?

mokagio commented:
Pushing this PR to the next milestone, 4.31, because I'm running the code-freeze today and this is labelled "enhancements, spring cleaning": it can safely wait till the next release.

eshurakov commented:
Hi @charliescheer, I believe you created this branch from `issue/1061-information-sheet-updates`. I updated base branch in this PR to `issue/1061-information-sheet-updates` so that we only see changes that you made on top of that branch.

eshurakov commented on Simplenote/Information/NoteInformationViewController.swift:
```diff
@@ -177,6 +181,24 @@ private extension NoteInformationViewController {
+        dragBar.translatesAutoresizingMaskIntoConstraints = false
+        NSLayoutConstraint.activate([
+            dragBar.topAnchor.constraint(equalTo: view.topAnchor, constant: 5),
+            dragBar.bottomAnchor.constraint(equalTo: view.topAnchor, constant: 10),
```
I think it's better to use `height`. Easier to understand dimensions :-)

charliescheer replied:
I have made those changes.  Check it out!  Let me know if you see anything else that needs to be updated!

eshurakov reviewed (changes requested):
Hey Charlie,

I think it's good to extract it in a separate class, not necessarily because we will reuse it later, but to keep view controller more tidy. 

Another thing is, we're using xib file to construct information screen, so let's move this handle to the xib and hide it in case we're presenting it in a popover. I just realised that checking for `UIDevice.isPad` won't always work, because in split view on iPad information screen is presented as a card if width class is compact or in a popover if width class is regular. What we can do is to check if there is a navigation controller attached to this view controller. When we present an information controller in a popover, we wrap it in a navigation controller, so if there is a navigation controller, we don't need to show a handle.

charliescheer committed:
Created SPDragBar class and file

charliescheer committed:
Changed creation of drag bar in the info sheet to use SPDragBar class

charliescheer committed:
Updated drag bar in info sheet to be inserted into the stackview

charliescheer committed:
added SPDragbar to xcode project

charliescheer committed:
add drag bar to XIB for information sheet

charliescheer committed:
Changed configureDragBar setup info sheet drag bar if not in a popover

jleandroperez commented on Simplenote/Information/NoteInformationViewController.swift:
```diff
@@ -177,6 +179,20 @@ private extension NoteInformationViewController {
+            ])
+        } else {
+            dragBar.isHidden = true
+        }
```
Since the initialization is already done in IB (`NoteInformationViewController`), would it be possible to move the Autolayout constraints there?

---

On a second note, we'd typically wanna avoid inline constants. Instead, you would probably wanna do something like this:

```
   dragBar.heightAnchor.constraint(equalToConstant: Metrics.dragBarSize.height),
   dragBar.widthAnchor.constraint(equalToConstant: Metrics.dragBarSize.width),
   dragBar.topAnchor.constraint(equalTo: view.topAnchor, constant: Metrics.dragBarTopPadding),
   dragBar.centerXAnchor.constraint(equalTo: view.centerXAnchor)
```

And at the bottom of the source file you could have:

```
private enum Metrics {
    static let dragBarSize = CGSize(width: 36, height: 5)
    static let dragBarTopPadding = CGFloat(5)
}
```

### Why:
Numbers used inline in the code are best avoided, for few reasons. It's harder to maintain, it's impossible to reuse. Suppose there were two components onscreen that needed the same **Top Padding**, without a constant, you'd have a `5` in two places.

**Plus:** if you had to change that `5`, without a proper constant you'd need to read the whole code, and patch two spots.

### And another note:
Why `enum Metrics` instead of, say, `struct Metrics`?  Because if you used `struct Metrics` this statement would be valid:

```
let metrics = Metrics()
```

It's a "workaround" we've been using everywhere, and allows us to namespace a bunch of constants. Other programming languages (such as C#) have a way to define a namespace that doesn't really belong to anything that can be instantiated. Swift... really doesn't, for now!

SylvesterWilmott commented:
Looks good!

Spacing and position is perfect. It appear in dark mode the handle remains dark, can we update the colour of the handle in dark mode to #ffffff with 0.2% opacity?

charliescheer committed:
Moved constraints for drag bar to interface builder

charliescheer committed:
Added dark mode support for the dragBar

charliescheer committed:
Removed unused else condition when configuring the dragbar

charliescheer commented:
Thanks for the suggestions @jleandroperez  and @SylvesterWilmott 

I have added dark mode for the drag bar
and moved the constraints to the IB.

Let me know if there are any more changes that need to be made

jleandroperez commented on Simplenote/Information/NoteInformationViewController.swift:
```diff
@@ -177,6 +179,12 @@ private extension NoteInformationViewController {
+    
+    func configureDragBar() {
+        if navigationController != nil {
+            dragBar.isHidden = true
```
Nipticky (cosmetic, feel free to disregard!!). You can simplify this one a bit:

```
dragBar.isHidden = navigationController != nil
```

charliescheer replied:
I like it, much cleaner

jleandroperez commented on Simplenote/SPDragBar.swift:
```diff
@@ -0,0 +1,24 @@
+        }
+        self.alpha = 0.2
+        self.layer.cornerRadius = 2.5
+        self.layer.masksToBounds = true
```
No need for `self` (unless you're inside a closure)

jleandroperez commented on Simplenote/SPDragBar.swift:
```diff
@@ -0,0 +1,24 @@
+        configure()
+    }
+    
+    func configure() {
```
This one should probably be `private`

jleandroperez commented on Simplenote/SPDragBar.swift:
```diff
@@ -0,0 +1,24 @@
+            self.backgroundColor = UIColor.white
+        } else {
+            self.backgroundColor = UIColor.black
+        }
```
This snippet can be simplified a bit:

```
backgroundColor = SPUserInterface.isDark ? .white : .black
```

charliescheer replied:
Much simpler!

charliescheer committed:
Updated configure method for SPDragBar

charliescheer committed:
Updated configureDragBar method for NoteInformationVC

charliescheer commented:
Thanks @jleandroperez  I made those changes

eshurakov commented on Simplenote/Information/NoteInformationViewController.swift:
```diff
@@ -177,6 +179,10 @@ private extension NoteInformationViewController {
         additionalSafeAreaInsets = UIEdgeInsets(top: headerStackView.frame.maxY, left: 0, bottom: 0, right: 0)
     }
+    
+    func configureDragBar() {
```
This can also be private :-)

eshurakov commented on Simplenote/Information/NoteInformationViewController.swift:
```diff
@@ -8,6 +8,7 @@ final class NoteInformationViewController: UIViewController {
     @IBOutlet private weak var screenTitleLabel: UILabel!
     @IBOutlet private weak var dismissButton: UIButton!
     @IBOutlet private weak var headerStackView: UIStackView!
+    @IBOutlet weak var dragBar: SPDragBar!
```
This can be private like other `@IBOutlet` properties

eshurakov commented on Simplenote/SPDragBar.swift:
```diff
@@ -0,0 +1,20 @@
+    }
+    
+    private func configure() {
+        backgroundColor = SPUserInterface.isDark ? .white : .black
```
Please take a look at `UIColor+Studio.swift` extension. You can define a pair of light / dark colors and use it here as a single UIColor. The benefit is that it will automatically update when theme changes. 
For example right now if you open an information sheet and change a theme (Simulator -> Features -> Toggle Appearance), the color of the handle won't update.

eshurakov reviewed (commented):
Hey Charlie, now that we show a handle instead of title and dismiss button, I believe we can safely delete the following things from the code and interface file:
* screenTitleLabel
* dismissButton
* headerStackView
* blurEffectView

SylvesterWilmott reviewed (approved):
Design looks good. Thanks @charliescheer

charliescheer committed:
Removed dismisButton from information sheet

charliescheer committed:
Removed titleLabel and associated methods from info sheet

charliescheer committed:
Remove blur affect and associated methods from info sheet

charliescheer committed:
Removed headerStackView and associated methods in info sheet

charliescheer committed:
Made dragbar methods private in informationSheetViewController

charliescheer committed:
Changed dragBar color to use ColorStudio colors/dark-light mode

charliescheer commented:
> now that we show a handle instead of title and dismiss button, I believe we can safely delete the following things from the code and interface file

Good call @eshurakov I have removed those extra things, and I don't see any issues with it having been removed.

I have also updated the dragbar to use the color studio colors and marked the methods as private.

Question for you though.  By removing the dismiss button we are also removing the accessibility label for dismissing.  Is there anything we can, or should do for accessibility labeling on the drag bar? maybe:
`    func configureAccessibility() {
        dragBar.accessibilityLabel = Localization.dismissAccessibilityLabel
    }`

@SylvesterWilmott can you take a quick look at the info sheet one more time and see if the spacing between the top of the info sheet and the first text item looks right to you?  It should have the correct spacing, but wanted to double check

eshurakov commented:
If we add a label to the drag bar, we also should handle tap on it so that it can be activated though accessibility. 
I think we can start by implementing [accessibilityPerformEscape](https://developer.apple.com/documentation/objectivec/nsobject/1615091-accessibilityperformescape) method to dismiss Info sheet. It is activated when you draw a “Z” with two fingers anywhere on the screen and should act like `ESC` key.

charliescheer committed:
Added accessibilityPerformEscape to NoteInformationSheet

charliescheer committed:
Changed informationVC drag bag to dismiss on tap if accessibility is on

charliescheer commented:
>If we add a label to the drag bar, we also should handle tap on it so that it can be activated though accessibility.
I think we can start by implementing accessibilityPerformEscape method to dismiss Info sheet. It is activated when you draw a “Z” with two fingers anywhere on the screen and should act like ESC key.

Great idea, I have added both of those things to the drag bar.  Both the Z to dismiss and the tap to dismiss should work with VoiceOver on

eshurakov commented on Simplenote/Information/NoteInformationViewController.swift:
```diff
@@ -121,76 +119,33 @@ private extension NoteInformationViewController {
-            screenTitleLabel.isHidden = true
-            dismissButton.isHidden = true
-        }
+        //TO DO: Should we do something for accessibility for dismissing the info sheet now that the dismiss button is gone?
```
I think can remove this one :-)

charliescheer replied:
Ah, missed that note.  Thanks!  Donoe

eshurakov reviewed (approved):
Looks good to me! 👍 

:shipit:

charliescheer committed:
Removed unused function and note for activating accessibility

charliescheer committed:
Removed call to configure accessibility method I removed

pachlava commented:
[testing] fix/feature verified on iPhone XR | iOS 14.3 | Simplenote 4.32.0.0
